### ISSUE TITLE
Revisit tile embedding training to help with Mario Maker PCG repo

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#192]
A modified version of block2vec was made in hopes of addressing this issue in the MarioMakerPCG repo: https://github.com/schrum2/MarioMakerPCG/issues/6

Although we're still struggling with that issue, I'm importing the modified code back here and testing it out.

**Comment by schrum2** [comment]
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/cd5b37be-2507-4ab1-8b73-cfe9d8e9bffd" />

The loss on the embedding model (with block2vec) is high, though nowhere near as high as in the MarioMakerPCG repo. This also plots the learning rate, but I should remove that. This code uses early stopping, so this was done in under 80 epochs.

The unconditional diffusion model that was trained based on it seems to produce good results. Here are samples from the last epoch:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/75e9db9e-4557-43f2-ab03-7a196ad346bf" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/4e346cce-c672-4cbb-8469-3a1b5dfa05af" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f1f5b951-9bc9-46ca-abe3-cb55df10454c" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/78ee79aa-c03c-4064-bd6d-9318623a82de" />

There is also a bit more variety in samples produced from run_diffusion afterwards:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/e5e05aef-8491-4252-a089-ca74485ee52d" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/bb1db321-1f98-434a-996e-0a454f1c4bb5" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/ca465e5f-797a-46ed-91d8-fa5b26aff6d5" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f7e9f102-ead8-456b-9304-ac0796e2bcce" />

These look great, though the embedding vector length is 16, which is actually higher than the tile count. Still, this seems to work with the unconditional model on Mario.

**Comment by schrum2** [comment]
Conditional model is still training, but I'm going to report results now since they look good. Here is the comparison of original samples to the model generated ones at epoch 160:

<img width="1321" height="894" alt="Image" src="https://github.com/user-attachments/assets/59057e40-6dd1-4e5f-b711-a46b14d11dd2" />

**Comment by schrum2** [comment]
Unconditional block2vec results with Mega Man were decent (issues with traversability, but look ok)

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/16170625-bbde-4488-90f1-f6cd938b3629" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/c3b4f4d5-f0fd-4ec6-a71c-3487cc015a99" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/58f11ae1-15a4-497c-bf05-6535df1edb18" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/861839e5-d56c-44aa-a2c8-81be4cf69414" />

embedding model stopped before 80 epochs. Stopped plotting learning rate with training loss, so here is what the embedding model's loss looked like:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/b040f79f-6ad8-4d16-aa80-2b2ffcc587a9" />

**Comment by schrum2** [comment]
When Mega Man embeddings are used to train a conditional model, the results are not that great:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/451fe9ff-a302-4ef0-84de-a19cb8085603" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/68d066f5-4692-4234-838f-83750b06545b" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/da2c60eb-74bc-4144-b3ec-da4e473185d2" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/9c75d6e4-083a-4530-840a-d4f7c55ff4af" />

Would be interesting to see how common this is, but I think it is comparable to what @fwilson12 was getting. Also, these block embeddings see to be using the MM-Simple set, so it would likely be worse with MM-Full.  We should really try training block embedding models on MM-Full, since that sort of demonstrates the challenge of scaling up the number of tiles.

I'll also try skipgrams to see if they are any better.

**Comment by schrum2** [comment]
Unconditional model with skipgram on Mario looks good:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/028266f6-8a53-44d7-a194-14c6800ab321" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/ba287438-41af-4d7f-a4f1-cd318e530350" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/37cdeda1-ed0d-4ca4-a843-3f0687f91beb" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/269c9201-4c0e-4e55-a670-76b164337a55" />

Run diffusion samples:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a02afc09-14bd-43bd-a398-d471832def25" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/3b6ddcfb-c11c-4e3b-8f67-de7d76d5de81" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/163b0249-4912-4471-b9bc-703ce2912808" />

**Comment by schrum2** [comment]
A conditional Mario model using skip grams seems fine:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/735109a7-ef58-4a0d-9f28-8f5bcc788b64" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a75af78a-46ad-40e5-a3fc-952219000a70" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f789f412-7af9-4c6c-8ea7-6d032f76e5f2" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/fddba35b-951d-4edc-837e-70e8d3630a74" />

**Comment by schrum2** [comment]
Currently training Mega Man models with skip grams. The skip gram model itself once again trained quickly due to early stopping:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/0d54de94-925f-44f7-9226-898c33502694" />
<img width="1000" height="400" alt="Image" src="https://github.com/user-attachments/assets/e5eeeb71-d33c-449f-b72f-d40b007ffe89" />

Interesting center tile distribution as well compared to Mario. This is the MM-simple tileset by the way. Still need to train on MM-full.

**Comment by schrum2** [comment]
Still training Mega Man with skipgrams at home (images above are for the "simple" tileset), but I've also started training on the "full" tileset, whose block2vec model reported this:

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/31caf665-aa33-42f5-9f5e-58f8796f666f" />
<img width="1000" height="400" alt="Image" src="https://github.com/user-attachments/assets/7d5d6a53-0213-4ec0-a150-37d7f63c3275" />

More low frequency tiles, which is closer to Mario Maker. Will be interesting to see what happens as a result.

**Comment by schrum2** [comment]
Diffusion definitely has a harder time with the MM-Full tileset. Note that these samples are from an unconditional diffusion model trained on block2vec embedding vectors of length 16 (contrast with 42 MM-Full tiles) trained on 3x3 windows.

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/56677aba-b7ed-43d3-ac52-31f712e21461" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a88ea59e-90f0-4637-b56a-cca6b304a360" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/2760f36d-91f4-4d30-b4af-efe06d90caf0" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/2d2e8658-22bd-417e-8775-dc1ffc0a8440" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/0b9ad02a-b640-436e-99b4-6d5985db01f5" />

The model seems to learn the structure of the walls and floor fairly well. It sometimes gets the ladders right. But all the little enemies and power-ups seem somewhat random, and in some ways inappropriate (multiple 1UPs in a single scene)

**Comment by schrum2** [comment]
Another bug was found, and the fix was copied over from MarioMakerPCG, so I need to redo a lot of these tests. However, the embeddings worked before ... hopefully they work even better after this fix. I'm mainly curious if the fix solves the problem with MM-Full, so I'm going to run this first:
```
MM_conditional-embeddings.bat 300 full block2vec 16 3
```
Note: I'm not sure skip grams work any more. This fix focused on block2vec

**Comment by schrum2** [comment]
Also, in my office, I'm training an unconditional model with the subsampling fix:
```
MM_unconditional-embeddings.bat 300 full block2vec 16 3
```
The tile sampling distribution is way different:

<img width="1000" height="400" alt="Image" src="https://github.com/user-attachments/assets/3b81a489-6e60-4b73-8662-b85e3f4ff424" />

**Comment by schrum2** [comment]
NOTE: These results are from BEFORE I fixed the subsampling approach. Still, I figured they were worh sharing, since they're not that bad. This is from a conditional model on the full Mega Man tile set. Comparin goriginal samples to generated ones.

<img width="960" height="703" alt="Image" src="https://github.com/user-attachments/assets/5e4aeb4c-fe76-4a18-a457-7ecc3764cc49" />

Seems a lot better than the unconditional results above. I think the text guidance helps a little with getting the right tiles into the scene, though there is still room for improvement. We'll see if the new block embedding model helps with that.

**Comment by schrum2** [comment]
Oh no. The newly fixed code for the embedding models actually produced worse results in Mega Man Full. This is an unconditional model:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/89cb36e9-e82a-4148-8844-5372e45432f9" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/232c4706-2573-465e-a6c0-d35c0b9da26d" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f2f51c7a-0339-44e9-abbd-90fd163aa43c" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/e92f6420-be8e-4762-92d3-271c55fee833" />

I guess having lower loss on the embedding doesn't necessarily mean better diffusion training. I should just be able to turn the subsampling off, though training with a different threshold tuned to Mega Man might work. Still need to check the skip grams too.

**Comment by schrum2** [comment]
Ok, I've thoroughly tested original Mario with a sub-sampling threshold of 0.05. Specifically:
- block2vec embeddings of length 8, unconditional diffusion. run_diffusion samples all look good. Plenty of enemies, pipes, coins, and stairs
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/c5326a26-0175-4378-b289-f727b917e50b" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/e25563c3-92bc-4a0e-ba91-ded9966d49e2" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/26d14aa6-bd50-45e0-9ed8-986412079da3" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/32a6986e-d613-4a05-b42f-05533f567454" />

- skipgram embeddings of length 8, unconditional diffusion. run_diffusion samples look good. Seem to favor undergraound levels, but coins, pipes, and enemies are present:
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/6e5fe397-3e13-40e5-94f9-6c7deb28b194" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/4d3e4642-5d00-4a8e-a4bc-e581ef8ba0df" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a76c0c29-4cd5-49e7-b043-6a9f9a39b792" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/01d130f7-c11d-4665-9d9c-bfb32ba2f5f4" />

- block2vec and skipgram with length 16, unconditional diffusion, also looks good (different bias in levels produced by run_diffusion ... I guess this is always a coin toss). Rare tiles are present ... though I guess I never saw cannons from any of the models.

- I Was training block2vec length 16 with conditional model, but got interrupted. However, the periodically generated samples looked just like the original training samples, and the caption score on the training data was above 0.8 by epoch 420.

I could do more verification on Mario with conditional models, but this seems to work fine. Different parameter settings have an impact, and I could just turn subsampling off to get what we had before, which also worked, but I think I'll stop testing Mario. Need to focus on Mega Man.

**Comment by schrum2** [comment]
I've been able to get decent unconditional diffusion results with block embeddings on MM-simple. Interetingly, these embeddings are of length 8. This might be helping, since 16 is too much capacity for only 13 tile types, but the increased use of negative sampling and other changes may also be helping. I'm also learning to understand the embedding analysis reports more. This one is supposedly good.

<img width="900" height="750" alt="Image" src="https://github.com/user-attachments/assets/0c5f8881-10dc-4b47-8ab7-7f5b5881a445" />

What makes this good is that some tiles are actually negatively correlated. If everything is positive, then it is harder to distinguish things, but this heatmap shows that some tiles are closely related and others are not, and being able to distinguish between tiles is important for using them as representations. Here are some unconditional samples from the diffusion model trained on this:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/9fd2e6c5-55ea-4d3f-b0d5-c71f7baf5956" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/d249f2d2-bfa6-4a69-9ebc-564c9e8931a4" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/4a00307b-f782-4edc-9a78-85cdc8daa544" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/8d3444bd-fe46-4c91-9fa6-f43aef322890" />

Not perfect, but within reach

**Comment by schrum2** [comment]
In contrast, training on MM-Full is harder. Here is the comparison heatmap:

<img width="922" height="922" alt="Image" src="https://github.com/user-attachments/assets/4d17979e-5f0b-4f2c-8bcc-c2fb5863d368" />

And some resulting samples:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/b4873050-4fcd-4938-9af6-d6424a0b51c8" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/18e286c1-af3a-4f39-af53-d9839d3a087c" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/3ad5b1a7-b23f-41ed-814e-804221047810" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/180bed1f-d898-4d63-8936-6f961566cca7" />

Lots of the rare tiles pop in randomly, and the heatmap is bad because it is all shades of red and no blue. I think that the embedding model is very sensitive to different parameters, and this one might actually need different subsampling because of how many more tiles types there are.

**Comment by schrum2** [comment]
Claude shared this useful insight:

"The real problem: extreme data sparsity, not center-frequency imbalance
Subsampling fixes one specific pathology: a tile dominating as a center so often that the gradient signal is swamped by it. Your simple tileset had that (tiles 0/1 at ~35–40% each), and subsampling correctly fixed it.
Your 41-tile dataset has a different and harder problem: 62% of your tiles (26 of 41) have under 10 dataset occurrences total, and several have just 1–3. Subsampling only ever removes samples from frequent centers — it can't manufacture context examples for a tile that appears twice in your whole dataset. No amount of threshold-tuning gives tile 16 (1 occurrence) or tile 0 (3 occurrences) enough distinct contexts to learn a meaningful direction. With vocab_size=41 and embedding_dim=16, you have 656 free parameters trying to be informed by, for many tiles, a single observed context window."

Basically, some tiles are just too rare in the dataset. The question is, what to do about it. Claude said increasing the negative samples or reducing the embedding dimension could help, but the thought that occurred to me (which Claude also suggested) was finding a way to increase the counts of the rare tiles. This either means that more data is needed, or that certain tiles need to be merged, as they already were in MM-Simple. The question is, which can we do without? I guess the trick is to figure out which ones are the rare ones first. This observation obviously applies directly to Mario Maker as well https://github.com/schrum2/MarioMakerPCG/issues/6 and there is another potential solution in both cases: train the block embedding models on WAY more data. The data exists, and even if we don't train the diffusion model on all of it, we can still train the embedding model. Something to consider.

**Comment by schrum2** [comment]
Analysis of MM-Full tiles that are super rare:
- 0: Starting spawn point. Should be replaced with empty space "-"/2 in training data.
- 23: Enemy? Why is this "?" in ASCII browser?
- 30: Another enemy. It appears along the bottom edge. Is this preventing it from being a "center" tile for embedding training?
- 22: super rare enemy. Why only 1 center appearance? 
- 15: 1UP power-up
- 18: Exit/goal power-up. Should be replaced with empty.
- 24: Another rare enemy (edge vs center?)
- 34: rare big-eye enemy
- 17: super rare power-up?
- 36: super rare enemy?
- 37: rare enemy?
- 38: rare enemy?

**Comment by schrum2** [comment]
Issue #197 is sort of a sub-issue of this. Essentially, embeddings work for Mario and MM-Simple, so all that is left to explore here is tile embeddings for MM-full.

One way of doing this is Issue #197 but the other is to try training a tile embedding model on a massive MMLV dataset, and see if it also has tile frequency problems.

**Comment by schrum2** [comment]
I just gave up on tile embeddings in Mario Maker: https://github.com/schrum2/MarioMakerPCG/issues/6

They can work (as shown above) but are difficult to configure, especially when there is sparse data. Perhaps fixing the tile set a bit could make a block embedding model work as good as one-hot, but it's not clear why we shouldn't just use one-hot instead. This issue revealed some issues with the Mega Man data that are being addressed elsewhere, but I actually don't think it is worth the effort of training a bunch of embeddings on Mega Man data to find the best one. We'll just use one-hot
